ion-schema/src/types.rs

```diff
@@ -241,41 +242,84 @@ impl Display for BuiltInTypeDefinition {
-        TypeDefinition::Named(TypeDefinitionImpl::new(
+    ) -> TypeDefinitionKind {
+        TypeDefinitionKind::Named(TypeDefinitionImpl::new(
             Some(name.into()),
             constraints.into(),
             None,
         ))
     }
 
     /// Creates an anonymous [`TypeDefinition`] using the [`IslConstraint`] defined within it
-    pub fn anonymous<A: Into<Vec<Constraint>>>(constraints: A) -> TypeDefinition {
-        TypeDefinition::Anonymous(TypeDefinitionImpl::new(None, constraints.into(), None))
+    pub fn anonymous<A: Into<Vec<Constraint>>>(constraints: A) -> TypeDefinitionKind {
+        TypeDefinitionKind::Anonymous(TypeDefinitionImpl::new(None, constraints.into(), None))
     }
 
     /// Provides the underlying constraints of [`TypeDefinitionImpl`]
     pub fn constraints(&self) -> &[Constraint] {
         match &self {
-            TypeDefinition::Named(named_type) => named_type.constraints(),
-            TypeDefinition::Anonymous(anonymous_type) => anonymous_type.constraints(),
+            TypeDefinitionKind::Named(named_type) => named_type.constraints(),
+            TypeDefinitionKind::Anonymous(anonymous_type) => anonymous_type.constraints(),
             _ => &[],
         }
     }
 
+    /// Provides the validation result for base built in type of the type definition with the given value
+    // This method is only used by nullable type reference for validation, it searches for a base type name which is built in type and nullable.
+    // It returns the result of validation for that nullable base type.
+    pub fn is_valid_for_base_nullable_type(
+        &self,
+        value: &IonSchemaElement,
+        type_store: &TypeStore,
+        ion_path: &mut IonPath,
+    ) -> bool {
+        // get a nullable built in base type name which can be used to perform validation to check for correct `null.*` type
+        let built_in_type_name = match self {
+            TypeDefinitionKind::Named(_) | TypeDefinitionKind::Anonymous(_) => {
+                // only built in type names are needed for base type
+                None
+            }
+            TypeDefinitionKind::BuiltIn(built_int_type) => match built_int_type {
+                BuiltInTypeDefinition::Atomic(ion_type, nullability) => {
+                    Some(format!("${ion_type}"))
+                }
+                BuiltInTypeDefinition::Derived(type_def) => {
+                    let type_name = type_def.name().as_ref().unwrap().to_string();
+                    if !type_name.starts_with('$') {
+                        Some(format!("${type_name}"))
+                    } else {
+                        Some(type_name)
+                    }
+                }
+            },
+        }
+        .unwrap(); // safe unwrap() because nullable type references are only used for built in types
+
+        // unwrap() here are safe because it has been verified while constructing the schema that built type definition exist in the type store
```

Comment:
This might not be safe if someone has tried to use `nullable::` on `document`, `nothing`, or anything derived from those types since there are no `$document` or `$nothing` types. Is there any way to construct a `nullable::document` or `nullable::nothing` type reference?

I took a look at the old function, and it did handle `nullable::nothing`. Is it intentional that we are getting rid of that?

Reply:
Programmatically only nullable types will be possible to generate based on Ion types for now using [`nullable_built_in_type_ref()`](https://github.com/amazon-ion/ion-schema-rust/pull/165/files#diff-6cf6ec7800b96bbfd4a54d935d0c1c4fd685aaa4322e866b059739d73ce32ce3R30)
Otherwise like you mentioned there is a check that happens while loading the schema to make sure nullable:document isn't constructed.
I will add `nullable:nothing` that was missed while working on this new change.